### PR TITLE
More robust Triangle-Triangle intersection

### PR DESCRIPTION
**Pull request by TheBlek**
- Replaced general case triangle-triangle intersection algorithm with [Moller's](https://web.stanford.edu/class/cs277/resources/papers/Moller1997b.pdf) 
- Added explicit handling of all corner cases (triangle degenerated into segments, points)
- Expanded triangle-triangle benchmark suite to test random triangles, including all corner cases. In future, it would be good to expand corner case testing with explicitly intersecting cases. Now because of the nature of segments, points and random numbers there are always not intersecting.
- Expanded tests a bit:
  - Added #655 as a test case
  - Added #762 as a test case
  - Added basic testing of corner cases

Benchmark results are looking good. 
Microbenchmarks of random intersections show improvements across the board. Although, I would not expect such a speedup in real world cases as you would not typically intersect thousands of triangles in a row.
Before:
Code_s41lv1rky7
After:
Code_MQZAMzC8s2

IntersectsGeometry benchmarks also show improvements. Much more so without BVH (perhaps because there are more triangle-triangle intersections).
Before:
Code_BvrQ4DtnRW
After:
Code_DNCbcq8cYf

Because benchmarks also changed on this branch, I needed to run them like this:
```
node .\benchmark\run-benchmark.js > .\new.md
git restore --source=upstream/master .\src\math\ExtendedTriangle.js
node .\benchmark\run-benchmark.js > .\old.md
```

Full test result output:
[new.md](https://github.com/user-attachments/files/21452113/new.md)
[old.md](https://github.com/user-attachments/files/21452114/old.md)

### CODE REVIEW

**Comment by gkjohnson** [comment]
Hey @TheBlek - thanks for the contribution and sorry for being slow on the response here. I've been preoccupied on other projects but I'll try to be more attentive here.

This triangle intersection code has been a significantly difficult problem and toying with epsilon values always seems like a losing battle (I've done my share of modifying this value, as well 😅). As with any adjustment like this my fear is that this will break other useful conditions. I'm wondering if there's a more robust triangle / triangle intersection function out there that we can use?

**Comment by gkjohnson** [comment]
I don't recall exactly where the current algorithm is from - but for the most part it was made with a focus on speed of execution. Though of course, at it turns out, these kinds of algorithms are particularly tricky. It would be best to start with a well-known, well-tested implementation & epsilons and deal with any other problems from there.

Unfortunately I don't have a specific algorithm or paper in mind to replace the algorithm with so any recommendations you have, like the one you've shared, are appreciated.

> since all those algorithms assume a generic case, all the degenerate cases should be handled beforehand.

I expect we're going to see issues no matter what but hopefully we'll still see an improvement. Regarding degenerate cases - if there are canonical ways to test for these then perhaps we can add this functionality to the beginning of the function or in a wrapper for the method.

Thanks for looking into this!

**Comment by TheBlek** [comment]
I've started working on explicitly handling degenerate cases before your comment and pushed results here if you want to check out how this might look anyway. I'm not particularly fond of the code I wrote but it gets the job done. And even the previously skipped test now passes.
Running the benchmark did not reveal any significant performance difference so this should be fine.
Nevertheless, I intend to check out a few algorithms from papers to see if they handle it better. Hence, draft status on the pull request.

**Comment by TheBlek** [comment]
I have implemented Moller's algorithm from [here](https://web.stanford.edu/class/cs277/resources/papers/Moller1997b.pdf). It handles corner cases for intersections pretty well, but still with quite a number of ifs. Also, I've found a couple optimizations that can be used regardless of main intersection algo:
1. ExtendedTriangle is constructing bounding sphere in .update() which seems to have never been used
2. Coplanar case contained checks with SAT axes that are cross products of triangles' edges. But that means that these axes were normals to the plane where the triangle lie. Hence they never would have been *separating* axes

For general case intersection implementation in this PR appears to outperform slightly against current implementation:
Before:
```json
{
  "name": "Math Functions",
  "results": [
    {
      "name": "IntersectTri w/o Target",
      "mean": 0.00026540000009117647,
      "median": 0.00029999998514540493,
      "min": 0.0001999999803956598,
      "max": 0.002099999983329326,
      "iterations": 1000
    },
    {
      "name": "IntersectTri w/ Target",
      "mean": 0.0005448999997170177,
      "median": 0.0005000000237487257,
      "min": 0.0004999999946448952,
      "max": 0.0011999999987892807,
      "iterations": 1000
    },
    {
      "name": "IntersectTri w/ Update",
      "mean": 0.000739700000412995,
      "median": 0.0007000000041443855,
      "min": 0.0005999999993946403,
      "max": 0.015699999989010394,
      "iterations": 1000
    }
  ]
},
```

After:
```json
{
  "name": "Math Functions",
  "results": [
    {
      "name": "IntersectTri w/o Target",
      "mean": 0.00020410000017727725,
      "median": 0.00020000000949949026,
      "min": 0.00009999997564591467,
      "max": 0.0009999999892897904,
      "iterations": 1000
    },
    {
      "name": "IntersectTri w/ Target",
      "mean": 0.0006590000002179295,
      "median": 0.0005999999993946403,
      "min": 0.0004999999946448952,
      "max": 0.02829999997629784,
      "iterations": 1000
    },
    {
      "name": "IntersectTri w/ Update",
      "mean": 0.0007584999999962747,
      "median": 0.0007999999797903001,
      "min": 0.000699999975040555,
      "max": 0.0027999999874737114,
      "iterations": 1000
    }
  ]
},
```
But benchmarking general case with one triangle pair intersected repeatedly seems not sufficient for evaluating algorithm performance. More about this later vvv

For more complex benchmarks results are inconclusive:
Before:
```json
{
  "name": "IntersectsGeometry w/ BVH",
  "mean": 2.469768100000394,
  "median": 2.4654999999911524,
  "min": 2.4144000000087544,
  "max": 3.7338999999919906,
  "iterations": 1000
},
{
  "name": "IntersectsGeometry w/o BVH",
  "mean": 38.769013178294834,
  "median": 38.8857999999891,
  "min": 37.89329999999609,
  "max": 46.952499999984866,
  "iterations": 387
},
{
  "name": "BVHCast",
  "mean": 58.34759186046406,
  "median": 58.29505000001518,
  "min": 57.19989999997779,
  "max": 66.85099999999511,
  "iterations": 258
}
```

After:
```json
{
  "name": "IntersectsGeometry w/ BVH",
  "mean": 1.6625149999998103,
  "median": 1.6542999999946915,
  "min": 1.596399999980349,
  "max": 2.8959000000031665,
  "iterations": 1000
},
{
  "name": "IntersectsGeometry w/o BVH",
  "mean": 78.6234848167548,
  "median": 78.59789999999339,
  "min": 78.13939999998547,
  "max": 79.39339999997173,
  "iterations": 191
},
{
  "name": "BVHCast",
  "mean": 42.55036628895265,
  "median": 42.52310000002035,
  "min": 41.87100000001374,
  "max": 45.99540000001434,
  "iterations": 353
}
```

I have tried to profile both implementations in order to figure out the inefficiency in `IntersectsGeometry w/o BVH`. The only difference that I've managed to find was that degenerate triangle are going into general case code path for current implementation and coplanar for my new one. (there appear to be [degenerate triangles in TorusGeometry](https://github.com/mrdoob/three.js/issues/30533))

I think it will be worthwhile to extend benchmarking code to intersect random triangles for each corner case e.g. random coplanar, random non-coplanar non-intersecting, random non-coplanar intersecting, degenerate triangles etc.

After that I think an informed choice can be made

**Comment by gkjohnson** [comment]
Amazing! Very nice work. Once it's cleaned up and ready for review I can take a look and give more detailed feedback.

> ExtendedTriangle is constructing bounding sphere in .update() which seems to have never been used

Yes sounds like there may be some vestigial code left over from previous iterations of different checks. Thanks for taking a deeper look.

> Coplanar case contained checks with SAT axes that are cross products of triangles' edges. But that means that these axes were normals to the plane where the triangle lie. Hence they never would have been separating axes

Can you explain this a bit more? The SAT axes code works by projecting points from both onto a given axis and checking of the min / max of the projected bounds are overlapping or not. If they're not they the shapes are not touching. In the case of the coplanar triangles the normal axis check should be able to check if the triangles are on the same plane (rather than having the same normal while on different planes - eg z = 0 and z = 1).

> For general case intersection implementation in this PR appears to outperform slightly against current implementation:

Amazing!

> For more complex benchmarks results are inconclusive:
> ...
> IntersectsGeometry w/ BVH

For these more complex cases it can be hard consider this a 1:1 comparsion. If there are cases where the new code is (correctly or incorrectly) detecting _more_ triangle intersections then the "intersectsGeometry" function may finish sooner or later. Basically it's not guaranteed that the exact same number of triangle-triangle intersections are being run.

> The only difference that I've managed to find was that degenerate triangle are going into general case code path for current implementation and coplanar for my new one. (there appear to be https://github.com/mrdoob/three.js/issues/30533)

We should probable check for these degenerate cases and decide what to do, as well. Normals and SAT axes will not be correct in these cases so it will need a separate strategy, I think. Regarding the torus case it's been useful to find this case. But perhaps we should consider changing the torus so degenerate faces are not present once this is figured out.

> I think it will be worthwhile to extend benchmarking code to intersect random triangles for each corner case e.g. random coplanar, random non-coplanar non-intersecting, random non-coplanar intersecting, degenerate triangles etc.
>
> After that I think an informed choice can be made

This sounds like a good plan. I really appreciate your thoroughness and rigor in working through. This is the kind of code that definitely needs that.

> P.S. Sorry for raw benchmarks with no comparison numbers, compare-bench-json.js gives me a cryptic error about incorrect json format in the file. Files themselves and the paths to them appear to be correct so I dunno what's wrong. Maybe you encountered something like that before?

I just ran the compare-bench and didn't see this issue. Could it have something to do with your platform? I'm running on Mac OSX.

**Comment by TheBlek** [comment]
> Can you explain this a bit more? The SAT axes code works by projecting points from both onto a given axis and checking of the min / max of the projected bounds are overlapping or not. If they're not they the shapes are not touching. In the case of the coplanar triangles the normal axis check should be able to check if the triangles are on the same plane (rather than having the same normal while on different planes - eg z = 0 and z = 1).

Yes. Since separating axes of triangles are plane normal and edges, we don't need to check all the cross products of those axes in a coplanar case. Cross products of triangles' edges are going to be collinear with the plane normal which we already checked. (In the new code normal axis can also be skipped since at that point we know triangles are coplanar).

Now that I think of it, not all cross products can be skipped either: we still need to check each `edge x normal`. Interestingly, tests pass even though my implementation skips those axes. I guess I will need to come up with a test that breaks my new implementation and fix the problem :)

> For these more complex cases it can be hard consider this a 1:1 comparsion. If there are cases where the new code is (correctly or incorrectly) detecting more triangle intersections then the "intersectsGeometry" function may finish sooner or later. Basically it's not guaranteed that the exact same number of triangle-triangle intersections are being run.

Thats a great point, I will check the exact intersection counts to see if thats the problem in the benchmark

> I just ran the compare-bench and didn't see this issue. Could it have something to do with your platform? I'm running on Mac OSX.

I'm running on Windows 11. I wouldn't be surprised if the paths are messed up on windows. I will investigate it more thoroughly then

**Comment by gkjohnson** [comment]
Hi @TheBlek I saw you added some new commits - is there anything you'd like me to take a look at?

**Comment by TheBlek** [comment]
Hi, yeah you can take a look at the new random benchmarks if you'd like. I've tried to stabilize the bench results but some margin for error still remains. For example, if you swap order of w/ and w/o target benchmarks, their relative performance also swaps. Maybe that's down to the machine configuration now, I don't know. Maybe your results will differ?

I've noticed weirdly high number of successful intersection for segment-point case, figured there is a bug in handling those special cases. I intend to look at this next, hopefully soon.

Sorry for dragging this out for so long

**Comment by gkjohnson** [comment]
> their relative performance also swaps. Maybe that's down to the machine configuration now, I don't know. Maybe your results will differ?

Yeah JS benchmarks seem to be tricky. May garbage collection etc is triggering causing differences. Or some times functions need to "warm up". But taking measurements at hundredths and thousandths of a second is bound to include some noise. Increasing the number of iterations per time sample to `let i = 2000;` seems to help stabalize the numbers a bit.

> Sorry for dragging this out for so long

No problem, of course! I appreciate any help on these projects 😁

**Comment by TheBlek** [comment]
Hi @gkjohnson I think I've done everything I wanted to here, so you can take a look.

Updated PR description to reflect everything done here.

Explicit handling takes a lot of lines of code, and I'm not sure how to do it better. Maybe you have any ideas?

File: src/math/ExtendedTriangle.js [hunk 1]
```diff
@@ -240,169 +544,183 @@ ExtendedTriangle.prototype.intersectsTriangle = ( function () {
...
-		} else {
+		}
 
-			// find the edge that intersects the other triangle plane
-			const count1 = triIntersectPlane( this, plane2, edge1 );
-			if ( count1 === 1 && other.containsPoint( edge1.end ) ) {
+		// How to do this more concisely?
+		let a1Proj, b1Proj, c1Proj;
+		let a2Proj, b2Proj, c2Proj;
+		switch ( componentIndex ) {
```
Comment (gkjohnson) [inline src/math/ExtendedTriangle.js:622]: I typically will do something like so:

```js
// global
const componentKeys = [ 'x', 'y', 'z' ];

// ...

const key = componentKeys[ componentIndex ];
a1Proj = this.a[ key ];
b1Proj = this.b[ key ];
c1Proj = this.c[ key ];
a2Proj = other.a[ key ];
b2Proj = other.b[ key ];
c2Proj = other.c[ key ];
```

File: benchmark/run-benchmark.js [hunk 2]
```diff
@@ -67,96 +70,245 @@ runSuiteWithOptions( '', { indirect: false } );
...
 
-} );
+// 	beforeAll( () => {
+
+// 		const geometry = new PlaneGeometry( 10, 10, 400, 400 );
+// 		const posAttr = geometry.getAttribute( 'position' );
+// 		for ( let x = 0; x <= 100; x ++ ) {
+
+// 			for ( let y = 0; y <= 100; y ++ ) {
+
```
[the remark this answers is not in the document]
Comment (TheBlek) [reply benchmark/run-benchmark.js]: No reason, it just was faster to bench triangles without it. Re-added

File: src/math/ExtendedTriangle.js [hunk 3]
```diff
@@ -132,89 +184,341 @@ ExtendedTriangle.prototype.closestPointToSegment = ( function () {
...
+		}
+
+		if ( target ) {
+
+			// TODO find two points that intersect on the edges and make that the result
+			if ( ! suppressLog ) {
+
+				// console.warn( 'ExtendedTriangle.intersectsTriangle: Triangles are coplanar which does not support an output edge. Setting edge to 0, 0, 0.' );
+
+			}
```
[the remark this answers is not in the document]
Comment (TheBlek) [reply src/math/ExtendedTriangle.js]: Not really. Commented that out to avoid logs. Returned back.

File: src/math/ExtendedTriangle.js [hunk 4]
```diff
@@ -132,89 +184,341 @@ ExtendedTriangle.prototype.closestPointToSegment = ( function () {
...
+	function coplanarIntersectsTriangle( self, other, target, suppressLog ) {
 
-		// find the edge that intersects the other triangle plane
-		const points = tri.points;
-		let count = 0;
-		let startPointIntersection = - 1;
-		for ( let i = 0; i < 3; i ++ ) {
+		// perform separating axis intersection test only for coplanar triangles
+		// There should be at least one non-degenerate triangle when calling this
+		// Otherwise we won't know the plane normal
```
[the remark this answers is not in the document]
Comment (TheBlek) [reply src/math/ExtendedTriangle.js]: This function is called only with at least one valid triangle. see intersectTriangleSegment and main intersectsTriangle function after handling degenerate cases.

File: src/math/ExtendedTriangle.js (hunk 1 above)
Comment (TheBlek) [reply src/math/ExtendedTriangle.js]: Looks great! Replaced.

**Comment by gkjohnson** [comment]
Thanks for the change! Realistically I think the details of this implementation are going to be too much for me to internalize but this is clearly a big improvement in terms of reliability and the performance seems at least on-par with the current one when integrating with bvh-csg so I'd like to merge it. I'll just need to ask for your support if there are issues or questions about the implementation later on, if that's okay.

It's something to deal with in the future but I'm also wondering if there's an existing, fast, well-tested implementation that could be ported already exists rather than implementing based on the paper 🤔

**Comment by TheBlek** [comment]
Thank you for the feedback!

> the performance seems at least on-par with the current one when integrating with bvh-csg

Great! Let me know if bvh-csg has issues with the triangle intersection part.

>  I'll just need to ask for your support if there are issues or questions about the implementation later on, if that's okay.

Of course! I'll be around.

> It's something to deal with in the future but I'm also wondering if there's an existing, fast, well-tested implementation that could be ported already exists rather than implementing based on the paper

To be frank, I did just port a reference C implementation of the paper :) With addition of explicit handling of the edge cases. I guess there could be intersection methods that do not require edge case handling, but I was unable to find such when doing research for this.

**Comment by gkjohnson** [comment]
> To be frank, I did just port a reference C implementation of the paper :)

Oh, great! I see now that you've included a reference to the c file in a comment.
